File: ion_crm_sales/ion_crm_sales/report/target_vs_actual/target_vs_actual.js
// Copyright (c) 2026, ard.ly and contributors
// For license information, please see license.txt

frappe.query_reports["Target vs Actual"] = {

	filters: [
		{
			fieldname: "quarter",
			label: __("Quarter"),
			fieldtype: "Select",
			reqd: 1,
			// Typical fiscal quarters. Adjust if you use custom quarters.
			options: [
				{label: __("Q1"), value: "Q1"},
				{label: __("Q2"), value: "Q2"},
				{label: __("Q3"), value: "Q3"},
				{label: __("Q4"), value: "Q4"},
			],
			// Default current quarter dynamically
			default: (function () {
				const m = parseInt(frappe.datetime.str_to_obj(frappe.datetime.get_today()).getMonth(), 10) + 1;
				if (m >= 1 && m <= 3) return "Q1";
				if (m >= 4 && m <= 6) return "Q2";
				if (m >= 7 && m <= 9) return "Q3";
				return "Q4";
			})(),
		},
		{
			fieldname: "fiscal_year",
			label: __("Fiscal Year"),
			fieldtype: "Link",
			options: "Fiscal Year",
			reqd: 1,
			// Try to default to current fiscal year if available
			default: frappe.datetime.get_today().split("-")[0], // simple default; adjust if FY != calendar year
			// Better: fetch current FY via query; uncomment if you want server-driven default
			// get_query: () => {
			//   return { filters: { disabled: 0 } };
			// },
		},
		{
			fieldname: "department",
			label: __("Department"),
			fieldtype: "Select",
			options: "Sales\nBusiness Accounts",
			reqd: 1,
			default: "Sales"
		},
	],
};

File: ion_crm_sales/ion_crm_sales/report/target_vs_actual/target_vs_actual.py
# Copyright (c) 2026, ard.ly and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    filters = frappe._dict(filters or {})

    parent_dt = "Sales Target and Commission Sheet"
    child_dt = "Commission Lines"

    conditions = [
        "stcs.quarter = %(quarter)s",
        "stcs.fiscal_year = %(fiscal_year)s",
        "stcs.status = 'Posted'"
    ]

    child_sales_person = "sales_person"
    child_target = "target_value"
    child_actual = "actual_sales"

    query = f"""
            SELECT
                line.`{child_sales_person}` AS sales_person,
                COALESCE(line.`{child_target}`, 0) AS target,
                COALESCE(line.`{child_actual}`, 0) AS actual,
                (COALESCE(line.`{child_actual}`, 0) - COALESCE(line.`{child_target}`, 0)) AS difference
            FROM `tab{parent_dt}` stcs
            JOIN `tab{child_dt}` line
              ON line.parent = stcs.name
            WHERE {" AND ".join(conditions)} AND line.department = %(department)s
            ORDER BY line.`{child_sales_person}`
        """

    data = frappe.db.sql(query, filters, as_dict=True)

    columns = [
        {"label": _("Sales Person"), "fieldname": "sales_person", "fieldtype": "Link", "options": "Sales Person",
         "width": 110},
        {"label": _("Target"), "fieldname": "target", "fieldtype": "Currency",
         "width": 160},
        {"label": _("Actual"), "fieldname": "actual", "fieldtype": "Currency",
         "width": 300},
        {"label": _("Difference"), "fieldname": "difference", "fieldtype": "Currency", "width": 120},
    ]
    return columns, data
